pkg/component/component.go

```diff
@@ -525,36 +530,39 @@ func ValidateComponentCreateRequest(client *occlient.Client, componentSettings c
 // ApplyConfig applies the component config onto component dc
 // Parameters:
 //	client: occlient instance
+//	kClient: kclient instance
 //	appName: Name of application of which the component is a part
 //	componentName: Name of the component which is being patched with config
 //	componentConfig: Component configuration
-//  	cmpExist: true if components exists in the cluster
+//	envSpecificInfo: Component environment specific information, available if uses devfile
+//  cmpExist: true if components exists in the cluster
 // Returns:
 //	err: Errors if any else nil
-func ApplyConfig(client *occlient.Client, componentConfig config.LocalConfigInfo, stdout io.Writer, cmpExist bool) (err error) {
-
-	// if component exist then only call the update function
-	if cmpExist {
-		if err = Update(client, componentConfig, componentConfig.GetSourceLocation(), stdout); err != nil {
-			return err
+func ApplyConfig(client *occlient.Client, kClient *kclient.Client, componentConfig config.LocalConfigInfo, envSpecificInfo envinfo.EnvSpecificInfo, stdout io.Writer, cmpExist bool) (err error) {
```

Comment:
We need to refactor this function. The URL list function makes a list call in checkIfURLChangesWillBeMade(), then in ApplyConfigCreateURL() (it calls Exists internally which calls list) and then in applyConfigDeleteURL(). Since list can be very expensive call, we should call list once and compute the URLs to be added/deleted and then proceed with the respective actions.